File: apps/web/src/lib/marketplace.ts
/**
 * Marketplace & Liquidity module — types and React Query hooks.
 */

import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query";
import { api } from "@/lib/api";

// ── Types ────────────────────────────────────────────────────────────────────

export type ListingType = "equity_sale" | "debt_sale" | "co_investment" | "carbon_credit";
export type ListingStatus =
  | "draft"
  | "active"
  | "under_negotiation"
  | "sold"
  | "withdrawn"
  | "expired";
export type ListingVisibility = "public" | "qualified_only" | "invite_only";
export type RFQStatus =
  | "submitted"
  | "under_review"
  | "countered"
  | "accepted"
  | "rejected"
  | "withdrawn";
export type TransactionStatus =
  | "pending"
  | "processing"
  | "completed"
  | "cancelled"
  | "disputed";

export interface ListingResponse {
  id: string;
  org_id: string;
  project_id: string | null;
  title: string;
  description: string;
  listing_type: ListingType;
  status: ListingStatus;
  visibility: ListingVisibility;
  asking_price: string | null;
  minimum_investment: string | null;
  currency: string;
  details: Record<string, unknown>;
  expires_at: string | null;
  project_name: string | null;
  project_type: string | null;
  geography_country: string | null;
  signal_score: number | null;
  rfq_count: number;
  created_at: string;
  updated_at: string;
}

export interface ListingListResponse {
  items: ListingResponse[];
  total: number;
}

export interface ListingCreateRequest {
  project_id?: string | null;
  title: string;
  description?: string;
  listing_type: ListingType;
  visibility?: ListingVisibility;
  asking_price?: number | null;
  minimum_investment?: number | null;
  currency?: string;
  details?: Record<string, unknown>;
  expires_at?: string | null;
}

export interface ListingUpdateRequest {
  title?: string;
  description?: string;
  visibility?: ListingVisibility;
  asking_price?: number | null;
  minimum_investment?: number | null;
  details?: Record<string, unknown>;
  expires_at?: string | null;
}

export interface RFQCreateRequest {
  proposed_price: number;
  currency?: string;
  message?: string;
  proposed_terms?: Record<string, unknown>;
}

export interface RFQRespondRequest {
  action: "accept" | "reject" | "counter";
  counter_price?: number | null;
  counter_terms?: Record<string, unknown>;
  message?: string;
}

export interface RFQResponse {
  id: string;
  listing_id: string;
  buyer_org_id: string;
  proposed_price: string;
  currency: string;
  status: RFQStatus;
  message: string;
  counter_price: string | null;
  counter_terms: Record<string, unknown> | null;
  submitted_by: string;
  listing_title: string | null;
  created_at: string;
  updated_at: string;
}

export interface RFQListResponse {
  items: RFQResponse[];
  total: number;
}

export interface TransactionResponse {
  id: string;
  listing_id: string;
  buyer_org_id: string;
  seller_org_id: string;
  rfq_id: string | null;
  amount: string;
  currency: string;
  status: TransactionStatus;
  terms: Record<string, unknown> | null;
  settlement_details: Record<string, unknown> | null;
  completed_at: string | null;
  listing_title: string | null;
  created_at: string;
  updated_at: string;
}

export interface TransactionListResponse {
  items: TransactionResponse[];
  total: number;
}

export interface PriceSuggestion {
  suggested_price: number;
  price_range_min: number;
  price_range_max: number;
  basis: string;
  comparable_count: number;
}

export interface ListingFilters {
  listing_type?: string;
  sector?: string;
  geography?: string;
  price_min?: number;
  price_max?: number;
  status?: string;
}

// ── Query Keys ───────────────────────────────────────────────────────────────

export const marketplaceKeys = {
  all: ["marketplace"] as const,
  listings: (filters?: ListingFilters) =>
    [...marketplaceKeys.all, "listings", filters ?? {}] as const,
  listing: (id: string) => [...marketplaceKeys.all, "listing", id] as const,
  sentRfqs: () => [...marketplaceKeys.all, "rfqs", "sent"] as const,
  receivedRfqs: () => [...marketplaceKeys.all, "rfqs", "received"] as const,
  transactions: () => [...marketplaceKeys.all, "transactions"] as const,
};

// ── Hooks ────────────────────────────────────────────────────────────────────

export function useListings(filters?: ListingFilters) {
  const params = new URLSearchParams();
  if (filters?.listing_type) params.set("listing_type", filters.listing_type);
  if (filters?.sector) params.set("sector", filters.sector);
  if (filters?.geography) params.set("geography", filters.geography);
  if (filters?.price_min != null) params.set("price_min", String(filters.price_min));
  if (filters?.price_max != null) params.set("price_max", String(filters.price_max));
  if (filters?.status) params.set("status", filters.status);
  const qs = params.toString();

  return useQuery({
    queryKey: marketplaceKeys.listings(filters),
    queryFn: () =>
      api
        .get<ListingListResponse>(`/marketplace/listings${qs ? `?${qs}` : ""}`)
        .then((r) => r.data),
  });
}

export function useListing(id: string | undefined) {
  return useQuery({
    queryKey: marketplaceKeys.listing(id ?? ""),
    queryFn: () =>
      api.get<ListingResponse>(`/marketplace/listings/${id}`).then((r) => r.data),
    enabled: !!id,
  });
}

export function useCreateListing() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (data: ListingCreateRequest) =>
      api.post<ListingResponse>("/marketplace/listings", data).then((r) => r.data),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: marketplaceKeys.listings() });
    },
  });
}

export function useUpdateListing() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      listingId,
      ...data
    }: ListingUpdateRequest & { listingId: string }) =>
      api
        .put<ListingResponse>(`/marketplace/listings/${listingId}`, data)
        .then((r) => r.data),
    onSuccess: (data) => {
      qc.invalidateQueries({ queryKey: marketplaceKeys.listing(data.id) });
      qc.invalidateQueries({ queryKey: marketplaceKeys.listings() });
    },
  });
}

export function useWithdrawListing() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (listingId: string) =>
      api.delete(`/marketplace/listings/${listingId}`),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: marketplaceKeys.listings() });
    },
  });
}

export function useSubmitRFQ() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      listingId,
      ...data
    }: RFQCreateRequest & { listingId: string }) =>
      api
        .post<RFQResponse>(`/marketplace/listings/${listingId}/rfq`, data)
        .then((r) => r.data),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: marketplaceKeys.sentRfqs() });
    },
  });
}

export function useSentRFQs() {
  return useQuery({
    queryKey: marketplaceKeys.sentRfqs(),
    queryFn: () =>
      api.get<RFQListResponse>("/marketplace/rfqs/sent").then((r) => r.data),
  });
}

export function useReceivedRFQs() {
  return useQuery({
    queryKey: marketplaceKeys.receivedRfqs(),
    queryFn: () =>
      api.get<RFQListResponse>("/marketplace/rfqs/received").then((r) => r.data),
  });
}

export function useRespondRFQ() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      rfqId,
      ...data
    }: RFQRespondRequest & { rfqId: string }) =>
      api
        .put<RFQResponse>(`/marketplace/rfqs/${rfqId}/respond`, data)
        .then((r) => r.data),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: marketplaceKeys.receivedRfqs() });
      qc.invalidateQueries({ queryKey: marketplaceKeys.listings() });
    },
  });
}

export function useTransactions() {
  return useQuery({
    queryKey: marketplaceKeys.transactions(),
    queryFn: () =>
      api
        .get<TransactionListResponse>("/marketplace/transactions")
        .then((r) => r.data),
  });
}

export function useCompleteTransaction() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (transactionId: string) =>
      api
        .post<TransactionResponse>(
          `/marketplace/transactions/${transactionId}/complete`,
          {}
        )
        .then((r) => r.data),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: marketplaceKeys.transactions() });
    },
  });
}

// ── Helpers ──────────────────────────────────────────────────────────────────

export const LISTING_TYPE_LABELS: Record<ListingType, string> = {
  equity_sale: "Equity Sale",
  debt_sale: "Debt Sale",
  co_investment: "Co-Investment",
  carbon_credit: "Carbon Credits",
};

export const RFQ_STATUS_LABELS: Record<RFQStatus, string> = {
  submitted: "Submitted",
  under_review: "Under Review",
  countered: "Countered",
  accepted: "Accepted",
  rejected: "Rejected",
  withdrawn: "Withdrawn",
};

export const TX_STATUS_LABELS: Record<TransactionStatus, string> = {
  pending: "Pending",
  processing: "Processing",
  completed: "Completed",
  cancelled: "Cancelled",
  disputed: "Disputed",
};

export function rfqStatusVariant(
  status: RFQStatus
): "success" | "error" | "warning" | "info" | "neutral" {
  if (status === "accepted") return "success";
  if (status === "rejected" || status === "withdrawn") return "error";
  if (status === "countered") return "warning";
  if (status === "submitted") return "info";
  return "neutral";
}

export function txStatusVariant(
  status: TransactionStatus
): "success" | "error" | "warning" | "info" | "neutral" {
  if (status === "completed") return "success";
  if (status === "cancelled" || status === "disputed") return "error";
  if (status === "processing") return "info";
  if (status === "pending") return "warning";
  return "neutral";
}

export function listingStatusVariant(
  status: ListingStatus
): "success" | "error" | "warning" | "info" | "neutral" {
  if (status === "sold") return "success";
  if (status === "withdrawn" || status === "expired") return "neutral";
  if (status === "under_negotiation") return "warning";
  if (status === "active") return "info";
  return "neutral";
}

export function formatPrice(value: string | null, currency: string): string {
  if (!value) return "—";
  const n = parseFloat(value);
  if (isNaN(n)) return "—";
  if (Math.abs(n) >= 1_000_000) return `${currency} ${(n / 1_000_000).toFixed(1)}M`;
  if (Math.abs(n) >= 1_000) return `${currency} ${(n / 1_000).toFixed(0)}K`;
  return `${currency} ${n.toFixed(2)}`;
}

File: apps/api/app/models/marketplace.py
"""Marketplace models: Listing, RFQ, Transaction."""

import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Any

from sqlalchemy import Date, ForeignKey, Index, Integer, Numeric, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import BaseModel
from app.models.enums import (
    ListingStatus,
    ListingType,
    ListingVisibility,
    RFQStatus,
    TransactionStatus,
)


class Listing(BaseModel):
    __tablename__ = "listings"
    __table_args__ = (
        Index("ix_listings_org_id", "org_id"),
        Index("ix_listings_project_id", "project_id"),
        Index("ix_listings_status", "status"),
        Index("ix_listings_listing_type", "listing_type"),
        Index("ix_listings_visibility", "visibility"),
        Index("ix_listings_org_id_status", "org_id", "status"),
    )

    org_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("organizations.id", ondelete="CASCADE"),
        nullable=False,
    )
    project_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("projects.id", ondelete="SET NULL"),
    )
    title: Mapped[str] = mapped_column(String(500), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False, default="")
    listing_type: Mapped[ListingType] = mapped_column(nullable=False)
    status: Mapped[ListingStatus] = mapped_column(
        nullable=False, default=ListingStatus.DRAFT
    )
    visibility: Mapped[ListingVisibility] = mapped_column(
        nullable=False, default=ListingVisibility.QUALIFIED_ONLY
    )
    asking_price: Mapped[Decimal | None] = mapped_column(Numeric(19, 4))
    minimum_investment: Mapped[Decimal | None] = mapped_column(Numeric(19, 4))
    currency: Mapped[str] = mapped_column(String(3), nullable=False, default="USD")
    details: Mapped[dict[str, Any] | None] = mapped_column(JSONB)
    expires_at: Mapped[date | None] = mapped_column(Date)

    # Relationships
    rfqs: Mapped[list["RFQ"]] = relationship(back_populates="listing")
    transactions: Mapped[list["Transaction"]] = relationship(back_populates="listing")

    def __repr__(self) -> str:
        return f"<Listing(id={self.id}, title={self.title!r}, status={self.status.value})>"


class RFQ(BaseModel):
    __tablename__ = "rfqs"
    __table_args__ = (
        Index("ix_rfqs_listing_id", "listing_id"),
        Index("ix_rfqs_buyer_org_id", "buyer_org_id"),
        Index("ix_rfqs_status", "status"),
        Index("ix_rfqs_buyer_status", "buyer_org_id", "status"),
    )

    listing_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("listings.id", ondelete="CASCADE"),
        nullable=False,
    )
    buyer_org_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("organizations.id", ondelete="CASCADE"),
        nullable=False,
    )
    proposed_price: Mapped[Decimal] = mapped_column(Numeric(19, 4), nullable=False)
    proposed_terms: Mapped[dict[str, Any] | None] = mapped_column(JSONB)
    currency: Mapped[str] = mapped_column(String(3), nullable=False, default="USD")
    status: Mapped[RFQStatus] = mapped_column(
        nullable=False, default=RFQStatus.SUBMITTED
    )
    message: Mapped[str] = mapped_column(Text, nullable=False, default="")
    counter_price: Mapped[Decimal | None] = mapped_column(Numeric(19, 4))
    counter_terms: Mapped[dict[str, Any] | None] = mapped_column(JSONB)
    submitted_by: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=False,
    )

    # Relationships
    listing: Mapped["Listing"] = relationship(back_populates="rfqs")

    def __repr__(self) -> str:
        return f"<RFQ(id={self.id}, status={self.status.value})>"


class Transaction(BaseModel):
    __tablename__ = "transactions"
    __table_args__ = (
        Index("ix_transactions_listing_id", "listing_id"),
        Index("ix_transactions_buyer_org_id", "buyer_org_id"),
        Index("ix_transactions_seller_org_id", "seller_org_id"),
        Index("ix_transactions_status", "status"),
    )

    listing_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("listings.id", ondelete="SET NULL"),
        nullable=False,
    )
    buyer_org_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("organizations.id", ondelete="CASCADE"),
        nullable=False,
    )
    seller_org_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("organizations.id", ondelete="CASCADE"),
        nullable=False,
    )
    rfq_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("rfqs.id", ondelete="SET NULL"),
    )
    amount: Mapped[Decimal] = mapped_column(Numeric(19, 4), nullable=False)
    currency: Mapped[str] = mapped_column(String(3), nullable=False, default="USD")
    status: Mapped[TransactionStatus] = mapped_column(
        nullable=False, default=TransactionStatus.PENDING
    )
    terms: Mapped[dict[str, Any] | None] = mapped_column(JSONB)
    settlement_details: Mapped[dict[str, Any] | None] = mapped_column(JSONB)
    completed_at: Mapped[datetime | None] = mapped_column()

    # Relationships
    listing: Mapped["Listing"] = relationship(back_populates="transactions")

    def __repr__(self) -> str:
        return f"<Transaction(id={self.id}, amount={self.amount}, status={self.status.value})>"
